cmd/endpoint.go

```diff
@@ -711,6 +711,13 @@ func CreateEndpoints(serverAddr string, foundLocal bool, args ...[]string) (Endp
 				return endpoints, setupType,
 					config.ErrInvalidErasureEndpoints(nil).Msg("all local endpoints should not have different hostnames/ips")
 			}
+			// Add missing port for local erasure setup.
+			for i := range endpoints {
+				_, _, err := net.SplitHostPort(endpoints[i].Host)
+				if err != nil {
+					endpoints[i].Host = net.JoinHostPort(endpoints[i].Host, serverAddrPort)
+				}
+			}
```

Comment:
why do we need this?

Reply:
Sorry I haven't update the description for pull request bcz that time it's too night in my timezone (but at least I marked it as WIP)

So, the story is: I deployed minio for production environment via helm chart with this config

```
drivesPerNode: <number of drive which is more than 4>
replicas: 1
pools: 3
```

then I got this error

```
Expected endpoint http://minio-0.minio-svc.<namespace>.svc.cluster.local/export-0, seen http://minio-0.minio-svc.<namespace>.svc.cluster.local:9000/export-0 
```

I look into the command, the command is like this

```
minio server \
    http://minio-{0...0}.minio-svc.<namespace>.svc.cluster.local/export-{0...<drivePerNode-1>} \
    http://minio-{1...1}.minio-svc.<namespace>.svc.cluster.local/export-{0...<drivePerNode-1>} \
    http://minio-{2...2}.minio-svc.<namespace>.svc.cluster.local/export-{0...<drivePerNode-1>} \
    -S /etc/minio/certs/ --address :9000 --console-address :9001
```

the error is from this line: https://github.com/minio/minio/blob/0e3bafcc54eedc25600cf6e3e6b666c905d96d6e/cmd/bootstrap-peer-server.go#L83

So here is the flow:
- The minio get the command and the arguments
- Then minio parse the argument
- Then in each of pods [this function](https://github.com/minio/minio/blob/edba7c987b2462c4014ed0150cd9339ec7c3f49e/cmd/endpoint.go#L572) is called 3 times based on the number of pools
- If the host in a pod is resolve to some IP, it will go to [this if branch](https://github.com/minio/minio/blob/edba7c987b2462c4014ed0150cd9339ec7c3f49e/cmd/endpoint.go#L707) which there is no information about the port
- because of that if branch, here is the endpoint configuration of each pod
pod minio-0
```
http://minio-0.minio-svc.<namespace>.svc.cluster.local/export-0
http://minio-0.minio-svc.<namespace>.svc.cluster.local/export-0
http://minio-0.minio-svc.<namespace>.svc.cluster.local/export-0
http://minio-0.minio-svc.<namespace>.svc.cluster.local/export-0
http://minio-1.minio-svc.<namespace>.svc.cluster.local:9000/export-0 
http://minio-1.minio-svc.<namespace>.svc.cluster.local:9000/export-0 
http://minio-1.minio-svc.<namespace>.svc.cluster.local:9000/export-0 
http://minio-1.minio-svc.<namespace>.svc.cluster.local:9000/export-0 
http://minio-2.minio-svc.<namespace>.svc.cluster.local:9000/export-0 
http://minio-2.minio-svc.<namespace>.svc.cluster.local:9000/export-0 
http://minio-2.minio-svc.<namespace>.svc.cluster.local:9000/export-0 
http://minio-2.minio-svc.<namespace>.svc.cluster.local:9000/export-0 
```

pod minio-1
```
http://minio-0.minio-svc.<namespace>.svc.cluster.local:9000/export-0 
http://minio-0.minio-svc.<namespace>.svc.cluster.local:9000/export-0 
http://minio-0.minio-svc.<namespace>.svc.cluster.local:9000/export-0 
http://minio-0.minio-svc.<namespace>.svc.cluster.local:9000/export-0 
http://minio-1.minio-svc.<namespace>.svc.cluster.local/export-0
http://minio-1.minio-svc.<namespace>.svc.cluster.local/export-0
http://minio-1.minio-svc.<namespace>.svc.cluster.local/export-0
http://minio-1.minio-svc.<namespace>.svc.cluster.local/export-0
http://minio-2.minio-svc.<namespace>.svc.cluster.local:9000/export-0 
http://minio-2.minio-svc.<namespace>.svc.cluster.local:9000/export-0 
http://minio-2.minio-svc.<namespace>.svc.cluster.local:9000/export-0 
http://minio-2.minio-svc.<namespace>.svc.cluster.local:9000/export-0 
```

pod minio-2
```
http://minio-0.minio-svc.<namespace>.svc.cluster.local:9000/export-0 
http://minio-0.minio-svc.<namespace>.svc.cluster.local:9000/export-0 
http://minio-0.minio-svc.<namespace>.svc.cluster.local:9000/export-0 
http://minio-0.minio-svc.<namespace>.svc.cluster.local:9000/export-0 
http://minio-1.minio-svc.<namespace>.svc.cluster.local:9000/export-0 
http://minio-1.minio-svc.<namespace>.svc.cluster.local:9000/export-0 
http://minio-1.minio-svc.<namespace>.svc.cluster.local:9000/export-0 
http://minio-1.minio-svc.<namespace>.svc.cluster.local:9000/export-0 
http://minio-2.minio-svc.<namespace>.svc.cluster.local/export-0
http://minio-2.minio-svc.<namespace>.svc.cluster.local/export-0
http://minio-2.minio-svc.<namespace>.svc.cluster.local/export-0
http://minio-2.minio-svc.<namespace>.svc.cluster.local/export-0
```

- Then this [line](https://github.com/minio/minio/blob/edba7c987b2462c4014ed0150cd9339ec7c3f49e/cmd/bootstrap-peer-server.go#L183) verify the configuration
- The verification is always fail bcz the configuration is different


So that line of code add the port information on each minio host in a pod which resolve to it's own pod, so the check is not fail.

Reply:
This is a bug your deployment @FwP-IDN there is no way that code will get triggered unless all your hosts resolve for 127.0.0.1

Reply:
> there is no way that code will get triggered unless all your hosts resolve for 127.0.0.1

you mean all hosts in a cluster?

Reply:
And btw the code is also triggered if the hosts resolve to one of hosts IP, **not always** 127.0.0.1

https://github.com/minio/minio/blob/7f214a0e46c24684fd76483f53b53bb31dfee5d3/cmd/net.go#L48-L78

Reply:
The problem is you have a setup issue the DNS needs to be resolved to unique nodes if they resolve to same hosts then it won't work @FwP-IDN . 

The fix you have is not correct can introduce regressions.

Reply:
This is the correct fix

```diff
git diff
diff --git a/cmd/endpoint.go b/cmd/endpoint.go
index 9a1f2ce9b..e7c4dc3f5 100644
--- a/cmd/endpoint.go
+++ b/cmd/endpoint.go
@@ -702,6 +702,17 @@ func CreateEndpoints(serverAddr string, foundLocal bool, args ...[]string) (Endp
                }
        }
 
+       // Add missing port in all endpoints.
+       for i := range endpoints {
+               _, port, err := net.SplitHostPort(endpoints[i].Host)
+               if err != nil {
+                       endpoints[i].Host = net.JoinHostPort(endpoints[i].Host, serverAddrPort)
+               } else if endpoints[i].IsLocal && serverAddrPort != port {
+                       // If endpoint is local, but port is different than serverAddrPort, then make it as remote.
+                       endpoints[i].IsLocal = false
+               }
+       }
+
        // All endpoints are pointing to local host
        if len(endpoints) == localEndpointCount {
                // If all endpoints have same port number, Just treat it as local erasure setup
@@ -718,17 +729,6 @@ func CreateEndpoints(serverAddr string, foundLocal bool, args ...[]string) (Endp
                // This means it is DistErasure setup.
        }
 
-       // Add missing port in all endpoints.
-       for i := range endpoints {
-               _, port, err := net.SplitHostPort(endpoints[i].Host)
-               if err != nil {
-                       endpoints[i].Host = net.JoinHostPort(endpoints[i].Host, serverAddrPort)
-               } else if endpoints[i].IsLocal && serverAddrPort != port {
-                       // If endpoint is local, but port is different than serverAddrPort, then make it as remote.
-                       endpoints[i].IsLocal = false
-               }
-       }
-
        uniqueArgs := set.NewStringSet()
        for _, endpoint := range endpoints {
                uniqueArgs.Add(endpoint.Host)
```

Reply:
okay done